federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,305 @@
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Core().Namespaces().List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Core().Namespaces().Watch(options)
+					},
+				},
+				&api_v1.Namespace{},
+				controller.NoResyncPeriodFunc(),
+				// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+				// would be just confirmation that some namespace opration suceeded.
+				util.NewTriggerOnMetaAndSpecChangesPreproc(
+					func(obj pkg_runtime.Object) { nc.deliverNamespaceObj(obj, nc.namespaceReviewDelay, false) },
+					func(obj pkg_runtime.Object) { util.SetClusterName(obj, cluster.Name) },
+				))
+		},
+
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the namespaces again.
+				nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(nc.clusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	nc.federatedUpdater = util.NewFederatedUpdater(nc.namespaceFederatedInformer,
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Create(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Update(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			err := client.Core().Namespaces().Delete(namespace.Name, &api.DeleteOptions{})
+			return err
+		})
+	return nc
+}
+
+func (nc *NamespaceController) Start() {
+	go nc.namespaceInformerController.Run(nc.stopChan)
+	nc.namespaceFederatedInformer.Start()
+	nc.namespaceDeliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		namespace := item.Value.(string)
+		nc.reconcileNamespace(namespace)
+	})
+	nc.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		nc.reconcileNamespacesOnClusterChange()
+	})
+	go func() {
+		select {
+		case <-time.After(time.Minute):
+			nc.namespaceBackoff.GC()
+		case <-nc.stopChan:
+			return
+		}
+	}()
+}
+
+func (nc *NamespaceController) Stop() {
+	nc.namespaceFederatedInformer.Stop()
+	close(nc.stopChan)
+}
+
+func (nc *NamespaceController) deliverNamespaceObj(obj interface{}, delay time.Duration, failed bool) {
+	namespace := obj.(*api_v1.Namespace)
+	nc.deliverNamespace(namespace.Name, delay, failed)
+}
+
+func (nc *NamespaceController) deliverNamespace(namespace string, delay time.Duration, failed bool) {
+	if failed {
+		nc.namespaceBackoff.Next(namespace, time.Now())
+		delay = delay + nc.namespaceBackoff.Get(namespace)
```

Comment:
Should this be "delay + backoff"? Why not just simply "backoff"? Can you please add a comment?